doc/src/devdocs/locks.md

```diff
@@ -3,158 +3,137 @@
->   * debuginfo_asyncsafe
->   * inference_timing_mutex
->   * ExecutionEngine::SessionLock
->
->     > flisp itself is already threadsafe, this lock only protects the `jl_ast_context_list_t` pool
->     > likewise, the ResourcePool<?>::mutexes just protect the associated resource pool
-
-The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
-
->   * global_roots_lock
->   * Module->lock
->   * JLDebuginfoPlugin::PluginMutex
->   * newly_inferred_mutex
-
-The following is a level 3 lock, which can only acquire level 1 or level 2 locks internally:
-
->   * Method->writelock
->   * typecache
-
-The following is a level 4 lock, which can only recurse to acquire level 1, 2, or 3 locks:
-
->   * MethodTable->writelock
-
-No Julia code may be called while holding a lock above this point.
-
-orc::ThreadSafeContext (TSCtx) locks occupy a special spot in the locking hierarchy. They are used to
-protect LLVM's global non-threadsafe state, but there may be an arbitrary number of them. By default,
-all of these locks may be treated as level 5 locks for the purposes of comparing with the rest of the
-hierarchy. Acquiring a TSCtx should only be done from the JIT's pool of TSCtx's, and all locks on
-that TSCtx should be released prior to returning it to the pool. If multiple TSCtx locks must be
-acquired at the same time (due to recursive compilation), then locks should be acquired in the order
-that the TSCtxs were borrowed from the pool.
-
-The following is a level 5 lock
-
->   * JuliaOJIT::EmissionMutex
-
-The following are a level 6 lock, which can only recurse to acquire locks at lower levels:
-
->   * jl_modules_mutex
-
-The following lock synchronizes IO operation. Be aware that doing any I/O (for example,
-printing warning messages or debug information) while holding any other lock listed above
-may result in pernicious and hard-to-find deadlocks. BE VERY CAREFUL!
-
->   * iolock
->   * Individual ThreadSynchronizers locks
->
->     > this may continue to be held after releasing the iolock, or acquired without it,
->     > but be very careful to never attempt to acquire the iolock while holding it
->
->   * Libdl.LazyLibrary lock
-
-The following is a level 7 lock, which can only be acquired when not holding any other locks:
-
->   * world_counter_lock
-
-
-The following is the root lock, meaning no other lock shall be held when trying to acquire it:
-
->   * toplevel
->
->     > this should be held while attempting a top-level action (such as making a new type or defining
->     > a new method): trying to obtain this lock inside a staged function will cause a deadlock condition!
->     >
->     >
->     > additionally, it's unclear if *any* code can safely run in parallel with an arbitrary toplevel
->     > expression, so it may require all threads to get to a safepoint first
-
-## Broken Locks
-
-The following locks are broken:
-
-  * toplevel
-
-    > doesn't exist right now
-    >
-    > fix: create it
```

Comment:
This also may deserve a call out still, since this synchronization for staticdata.c does exist now, but is quite unusual: there is a call to a function called `jl_safepoint_suspend_all_threads`, which causes all other threads to block at their next safepoint/safe-region until that lock is released. It therefore sits at about level 2 of the hierarchy, in that no lock is permitted to be acquired inside, unless that lock disallows having safepoints (which is level 1).